yaml.go

```diff
@@ -11,20 +11,98 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
-// Marshal marshals the object into JSON then converts JSON to YAML and returns the
-// YAML.
-func Marshal(o interface{}) ([]byte, error) {
-	j, err := json.Marshal(o)
-	if err != nil {
-		return nil, fmt.Errorf("error marshaling into JSON: %v", err)
+// An Encoder serializes values into YAML and writes them using its io.Writer.
+type Encoder struct {
+	w io.Writer
+}
+
+// NewEncoder creates a new Encoder.
+func NewEncoder(w io.Writer) *Encoder {
+	return &Encoder{w: w}
+}
+
+// Encode serializes a value into YAML and writes it using the Encoder's
+// io.Writer.
+func (e *Encoder) Encode(o interface{}) error {
+	pipeReader, pipeWriter := io.Pipe()
+	defer pipeReader.Close()
+
+	var intermediate interface{}
+	jsonEnc := json.NewEncoder(pipeWriter)
+	// We are using yaml.Decoder here (instead of json.Decoder) because the Go
+	// JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}; it always picks float64.
+	// go-yaml preserves the number type.
+	yamlDec := yaml.NewDecoder(pipeReader)
+	yamlEnc := yaml.NewEncoder(e.w)
+
+	var jsonErr error
+	go func() {
```

Comment:
spawning goroutines here and in Decode is a little hard to reason about... a few things I noticed/wondered

* if this panics, the calling program will exit instead of propagating the panic to the Encode() caller... I'd expect the panic to propagate instead
* is it better to start the json encoder async and the yaml decoder inline or vice-versa?
* is it better to leave the pipereader unbuffered or wrap in a buffer?

Reply:
FWIW, this is adapted from the [official `io.Pipe` docs](https://pkg.go.dev/io#example-Pipe).